components/region_cache_memory_engine/src/background.rs

```diff
@@ -451,6 +603,76 @@ impl BackgroundRunnerCore {
         #[cfg(test)]
         flush_epoch();
     }
+
+    fn evict_on_soft_limit_reached(&mut self) {
+        if self.range_stats_manager.is_none() {
+            warn!("range stats manager is not initialized, cannot evict on soft limit reached");
+            return;
+        }
+        let range_stats_manager = self.range_stats_manager.as_ref().unwrap();
+        let to_shrink_by = self
+            .memory_controller
+            .mem_usage()
+            .checked_sub(self.memory_controller.soft_limit_threshold());
+        if to_shrink_by.is_none() {
+            return;
+        }
+        let mut remaining = to_shrink_by.unwrap();
+        let mut ranges_to_evict = Vec::<(CacheRange, u64)>::with_capacity(256);
+        range_stats_manager.collect_candidates_for_eviction(&mut ranges_to_evict);
+        let ranges_to_evict = ranges_to_evict
+            .iter()
+            .filter(|(range, _)| self.engine.read().range_manager().contains_range(range))
+            .rev();
+        for (range, approx_size) in ranges_to_evict {
+            if remaining == 0 {
+                break;
+            }
+            if self.engine.write().mut_range_manager().evict_range(range) {
```

Comment:
TODO: I need to add a check to see if this was recently loaded. It should be easy here.

Reply:
Should call engine.evict_range directly as it needs to do some cleanup work.

Reply:
> Should call engine.evict_range directly as it will do some cleanup works

@afeinberg

Reply:
`self.engine.write().mut_range_manager().evict_range(range)` returns true means the range can be deleted from the skiplist directly and thus not need to wait for the drop of some snapshots. So we have to delete the range if it return true, and this logic is already have in RangeCacheMemoryEngine::evict_range. It seems we cannot get the engine here, so we have to delete the range mannually.

Reply:
Thanks for bringing this to my attention. Will do.

Reply:
@SpadeA-Tang Done. Is what I am doing now what you had in mind?